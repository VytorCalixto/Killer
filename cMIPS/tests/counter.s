# Testing the internal counter is difficult because it counts clock cycles
# rather than instructions -- if the I/O or memory latencies change then
# the simulation output also changes and comparisons are impossible.
# To perform comparisons the trick is to only update the counter when the
# PC is updated, thus counting instructions.  If there are any stalls,
# the counter also stalls and the simulation outputs are the same
# regardless of the relative latencies.

	# mips-as -O0 -EL -mips32r2
	.include "cMIPS.s"
	.text
	.align 2
	.set noat
	.set noreorder
	.global _start
	.global _exit
	.global exit
	.ent _start
_start: nop
        li   $k0, cop0_STATUS_reset # RESET, kernel mode, all else disabled
        mtc0 $k0, cop0_STATUS
	li   $sp,(x_DATA_BASE_ADDR+x_DATA_MEM_SZ-8) # initialize SP: ramTop-8
        li   $k0, 0x1800ff01  # RESET_STATUS, kernel mode, interr enabled
        mtc0 $k0, cop0_STATUS
        li   $k0, cop0_CAUSE_reset  # RESET, disable counter
        mtc0 $k0, cop0_CAUSE

	la   $15,x_IO_BASE_ADDR
	nop
	jal main
	nop
exit:	
_exit:	nop	# flush pipeline
	nop
	nop
	nop
	nop
	wait	# then stop VHDL simulation
	nop
	nop
	.end _start
	
	.org x_EXCEPTION_0180,0
	.global _excp_180
	.global excp_180
	.ent _excp_180
excp_180:	
_excp_180:
        mfc0  $k0,cop0_CAUSE	# read CAUSE and
	sw    $k0,0($15)       	#   print it
	eret
	.end _excp_180

	
	.org x_EXCEPTION_0200,0
	.ent _excp_200
excp_200:	
_excp_200:	
	sw    $13, x_IO_ADDR_RANGE($15) # blank line
        mfc0  $k0, cop0_CAUSE	# read CAUSE and
	sw    $k0, 0($15)      	#   print it
	mfc0  $k1, cop0_COUNT  	# print current COUNT
	sw    $k1, 0($15)
	addiu $k1, $k1,(64-9) 	# interrupt again in approx 64 cycles
	mtc0  $k1, cop0_COMPARE
	sw    $k1, 4($15)      	# show new limit

	li   $k0, 0x1800ff01   	# enable interrupts
        mtc0 $k0, cop0_STATUS
	sw   $13, x_IO_ADDR_RANGE($15) # blank line
	eret
	.end _excp_200


	.org x_ENTRY_POINT,0
main:	la    $15, x_IO_BASE_ADDR
 	li    $13, '\n'

	addiu $5,$zero,(64-9)   # interrupt again in approx 64 cycles
	mtc0  $5,cop0_COMPARE

	mfc0  $5,cop0_CAUSE
	li    $6,0xf7ffffff    	# CAUSE(DisableCount) <= 0
	and   $5,$5,$6
	mtc0  $5,cop0_CAUSE   	# enable counter

	addiu $11,$12,1        	# this is a NOP

here:	addiu $11,$12,2	        # this is a NOP
	mfc0  $16,cop0_COUNT   	# print current COUNT
	sw    $16,0($15)
	slti  $1,$16,0x200     	# COUNT > 0x200 => stop counter and program
	beq   $1,$zero,there
	addiu $11,$12,3		# this is a NOP
	addiu $11,$12,4		# this is a NOP
	b here
	addiu $11,$12,5		# this is a NOP

	sw    $13, x_IO_ADDR_RANGE($15)
there:	mfc0  $5,cop0_CAUSE
	lui   $6,0x0880        	# CAUSE(DisableCount) <= 1
	or    $5,$5,$6
	mtc0  $5,cop0_CAUSE   	# disable counter
	addiu $11,$12,6		# this is a NOP
	mfc0  $16,cop0_COUNT   	# print current COUNT
	sw    $16,0($15)
	addiu $11,$12,7		# this is a NOP
	addiu $11,$12,8		# this is a NOP
	addiu $11,$12,9		# this is a NOP
	addiu $11,$12,10	# this is a NOP
	mfc0  $16,cop0_COUNT   	# print current COUNT
	sw    $16,0($15)
	j exit
	nop
