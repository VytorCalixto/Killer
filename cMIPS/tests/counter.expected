0000001e
00000027
00000030

40808000
0000003f
00000076

00000039
00000053
0000005c
00000065
0000006e

40808000
0000007d
000000b4

00000077
00000091
0000009a
000000a3
000000ac

40808000
000000bb
000000f2

000000b5
000000cf
000000d8
000000e1
000000ea

40808000
000000f9
00000130

000000f3
0000010d
00000116
0000011f
00000128

40808000
00000137
0000016e

00000131
0000014b
00000154
0000015d
00000166

40808000
00000175
000001ac

0000016f
00000189
00000192
0000019b
000001a4

40808000
000001b3
000001ea

000001ad
000001c7
000001d0
000001d9
000001e2

40808000
000001f1
00000228

000001eb
00000205
0000020f
00000215
